File: LoginRadiusSDK/LoginRadiusSocialLoginManager.m
//
//  LoginRadiusSocialLoginManager.m
//
//  Copyright © 2016 LoginRadius Inc. All rights reserved.
//

#import "LoginRadiusSocialLoginManager.h"
#import "LoginRadiusTwitterLogin.h"
#import "LoginRadiusFacebookLogin.h"
#import "LoginRadiusWebLoginViewController.h"
#import "LoginRadiusSafariLogin.h"
#import <SafariServices/SafariServices.h>
#import "LoginRadiusREST.h"
#import "LoginRadius.h"

@interface LoginRadiusSocialLoginManager() {}
@property(nonatomic, strong) LoginRadiusTwitterLogin * twitterLogin;
@property(nonatomic, strong) LoginRadiusFacebookLogin * facebookLogin;
@property(nonatomic, strong) LoginRadiusSafariLogin * safariLogin;
@property (assign, readonly, nonatomic) BOOL isSafariLogin;
@property (assign, readonly, nonatomic) BOOL isFacebookNativeLogin;
@property BOOL EmailVerified;
@property BOOL PhoneIdVerified;
@end

@implementation LoginRadiusSocialLoginManager

+ (instancetype)sharedInstance {
    static dispatch_once_t onceToken;
    static LoginRadiusSocialLoginManager *instance;

    dispatch_once(&onceToken, ^{
        instance = [[LoginRadiusSocialLoginManager alloc] init];
    });

    return instance;
}

-(instancetype)init {
    self = [super init];
    if (self) {
        _twitterLogin = [[LoginRadiusTwitterLogin alloc] init];
        _facebookLogin = [[LoginRadiusFacebookLogin alloc] init];
        _safariLogin = [[LoginRadiusSafariLogin alloc] init];
    }
    return self;
}

#pragma mark Login Methods
-(void)loginWithProvider:(NSString*)provider
            inController:(UIViewController*)controller
       completionHandler:(LRAPIResponseHandler)handler {
   
    // Use SFSafariViewController if available by defualt. Recommended approach
    if ([SFSafariViewController class] != nil) {
        _isSafariLogin = YES;
        [self.safariLogin loginWithProvider:provider inController:controller completionHandler:handler];
        return;
    }

    // Use web login
    LoginRadiusWebLoginViewController *webVC = [[LoginRadiusWebLoginViewController alloc] initWithProvider:provider completionHandler:handler];
    UINavigationController *navVC = [[UINavigationController alloc] initWithRootViewController:webVC];
    [controller presentViewController:navVC animated:YES completion:nil];
}


- (void)nativeFacebookLoginWithPermissions:(NSDictionary *)params
withSocialAppName:(NSString *)socialAppName
                              inController:(UIViewController *)controller
completionHandler:(LRAPIResponseHandler)handler {
   // self.handler=handler;
    _isFacebookNativeLogin = YES;
    [self.facebookLogin loginfromViewController:controller parameters:params withSocialAppName:socialAppName handler:handler];
}

- (void)convertTwitterTokenToLRToken:(NSString *)twitterAccessToken twitterSecret:(NSString *)twitterSecret
                         withSocialAppName:(NSString *)socialAppName
                        inController:(UIViewController *)controller completionHandler:(LRAPIResponseHandler)handler {
    [self.twitterLogin getLRTokenWithTwitterToken:twitterAccessToken twitterSecret:twitterSecret withSocialAppName:socialAppName inController:controller handler:handler];
    
}

- (void)convertGoogleTokenToLRToken:(NSString *)google_token google_refresh_token:(NSString *)google_refresh_token google_client_id:(NSString *)google_client_id
                        withSocialAppName:(NSString *)socialAppName
                       inController:(UIViewController *)controller completionHandler:(LRAPIResponseHandler)handler {
    NSString *refresh_token = google_refresh_token ? google_refresh_token: @"";
    NSString *client_id = google_client_id ? google_client_id: @"";
    NSMutableDictionary *dictParam = [NSMutableDictionary dictionaryWithDictionary:@{@"key": [LoginRadiusSDK apiKey],
    @"google_access_token" : google_token,
    @"refresh_token" : refresh_token,
    @"client_id" : client_id
    }];
    if(socialAppName && [socialAppName length]) {
        [dictParam setValue:socialAppName forKey:@"socialappname"];
    }
    
    [[LoginRadiusREST apiInstance] sendGET:@"api/v2/access_token/google"
                               queryParams:dictParam
                         completionHandler:^(NSDictionary *data, NSError *error) {

         handler(data, error);
        
    }];

}

- (void)convertWeChatCodeToLRToken:(NSString *)code completionHandler:(LRAPIResponseHandler)handler {
    [[LoginRadiusREST apiInstance] sendGET:@"api/v2/access_token/wechat"
                            queryParams:@{ @"key": [LoginRadiusSDK apiKey],
                                           @"code" : code
                                         }
                         completionHandler:^(NSDictionary *data, NSError *error) {
         handler(data, error);
        
    }];

}

- (void)convertAppleCodeToLRToken:(NSString *)code
                 withSocialAppName:(NSString *)socialAppName
                completionHandler:(LRAPIResponseHandler)handler {
    NSMutableDictionary *dictParam = [NSMutableDictionary dictionaryWithDictionary:@{ @"key": [LoginRadiusSDK apiKey],
      @"code" : code
    }];
   if(socialAppName && [socialAppName length]) {
        [dictParam setValue:socialAppName forKey:@"socialappname"];
    }
    
    [[LoginRadiusREST apiInstance] sendGET:@"api/v2/access_token/apple"
                            queryParams:dictParam
                         completionHandler:^(NSDictionary *data, NSError *error) {
         handler(data, error);
        
    }];

}


- (void)logout {
    // Only facebook native login stores sessions that we have to clear
    [self.facebookLogin logout];
}

#pragma mark Application delegate methods
- (void)applicationLaunchedWithOptions:(NSDictionary *)launchOptions {
    [self.facebookLogin applicationLaunchedWithOptions:launchOptions];
}

- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {

    BOOL canOpen = NO;


    if (_isFacebookNativeLogin)
    {
        _isFacebookNativeLogin = NO;
        canOpen = [self.facebookLogin application:application openURL:url sourceApplication:sourceApplication annotation:annotation];
    } else if (_safariLogin)
    {
        _isSafariLogin = NO;
        canOpen = [self.safariLogin application:application openURL:url sourceApplication:sourceApplication annotation:annotation];
    }

    return canOpen;
}


@end

File: Sources/LoginRadiusSocialLoginManager.swift
//
//  LoginRadiusSocialLoginManager.swift
//  LoginRadiusSwift SDK
//
//  Created by Megha Agarwal.
//  Copyright © 2023 LoginRadius Inc. All rights reserved.
//

import Foundation
import SafariServices

/**
 *  Social Login Manager
 */


public class LoginRadiusSocialLoginManager {
    public static let sharedInstance = LoginRadiusSocialLoginManager()
    
    // MARK: - Init
    
    public var twitterLogin = LoginRadiusTwitterLogin()
    public var facebookLogin = LoginRadiusFacebookLogin()
    public var safariLogin = LoginRadiusSafariLogin()
    
    private(set) var isSafariLogin: Bool = false
    
    public var isFacebookNativeLogin = false
    
    public init() {
        twitterLogin = LoginRadiusTwitterLogin()
        facebookLogin = LoginRadiusFacebookLogin()
        safariLogin = LoginRadiusSafariLogin()
    }
    
    // MARK: -  Methods
    
    /**
     *  Login with the given provider
     *
     *  @param provider   provider name in small case (e.g facebook, twitter, google, linkedin, yahoo etc)
     *  @param controller view controller where social login take place should not be nil
     *  @param handler    code block executed after completion
     */
   
    
    
    public func login(withProvider provider: String,
                      inController controller: UIViewController,
                      completionHandler handler: @escaping LRAPIResponseHandler) {
        
        // Use SFSafariViewController if available by default. Recommended approach
        if SFSafariViewController.self != nil {
            self.isSafariLogin = true
            self.safariLogin.login(withProvider: provider, inController: controller, completionHandler: handler)
            return
        }
        
        //Use web login
        let webVC = LoginRadiusWebLoginViewController(provider: provider, completionHandler: handler)
        let navVC = UINavigationController(rootViewController: webVC)
        controller.present(navVC, animated: true, completion: nil)
    }
    
    /**
     *  Native Facebook Login
     *
     *  @param params     dict of parameters
     These are the valid keys
     - facebookPermissions : should be an array of strings
     - facebookLoginBehavior : New FB SDK is supporting FBSDKLoginBehaviorBrowser only
     recommended approach is to use FBSDKLoginBehaviorSystemAccount
     *  @param socialAppName  should have unique social app name as a provider in case of multiple social apps for the same provider (eg. facebook_<social app name> )
     *  @param controller view controller where social login take place should not be nil
     *  @param handler    code block executed after completion
     */
    public func nativeFacebookLoginWithPermissions(params: [String: Any], withSocialAppName socialAppName: String, inController controller: UIViewController, completionHandler handler: @escaping LRAPIResponseHandler) {
        self.isFacebookNativeLogin = true
        self.facebookLogin.login(from: controller, parameters: params, withSocialAppName: socialAppName, handler: handler)
    }
    
    /**
     Native Twitter Login
     
     @param twitterAccessToken Your Twitter User's Access Token
     @param twitterSecret Your Twitter app's Consumer Secret
     @param socialAppName  should have unique social app name as a provider in case of multiple social apps for the same provider (eg. twitter_<social app name> )
     @param controller view controller where Twitter login take place should not be nil
     @param handler code block executed after completion
     */
    public func convertTwitterTokenToLRToken(twitterAccessToken: String, twitterSecret: String, withSocialAppName socialAppName: String, inController controller: UIViewController, completionHandler handler: @escaping LRAPIResponseHandler) {
        self.twitterLogin.getLRToken(withTwitterToken: twitterAccessToken, twitterSecret: twitterSecret, socialAppName: socialAppName, inController: controller, handler: handler)
    }
    
    /**
     Native Google Login
     
     @param google_token Google Access Token
     @param google_refresh_token Google refresh_token
     @param google_client_id Google client_id
     @param socialAppName  should have unique social app name as a provider in case of multiple social apps for the same provider (eg. google_<social app name> )
     @param controller view controller where google login take place should not be nil
     @param handler code block executed after completion
     */
    public func convertGoogleTokenToLRToken(googleToken: String, googleRefreshToken: String, googleClientId: String, withSocialAppName socialAppName: String?, inController controller: UIViewController, completionHandler handler: @escaping LRAPIResponseHandler) {
        let refreshToken = googleRefreshToken ?? ""
        let clientId = googleClientId ?? ""
        var dictParam: [String: Any] = [
            "key": LoginRadiusSDK.sharedInstance.apiKey,
            "google_access_token": googleToken,
            "refresh_token": refreshToken,
            "client_id": clientId
        ]
        
        if let socialAppName = socialAppName, !socialAppName.isEmpty {
            dictParam["socialappname"] = socialAppName
        }
        
        LoginRadiusREST.apiInstance.sendGET(url: "api/v2/access_token/google", queryParams: dictParam) { (data, error) in
            handler(data, error)
        }
    }
    
    /**
     Native WeChat Login
     
     @param code WeChat Access Code
     @param handler code block executed after completion
     */
    public func convertWeChatCodeToLRToken( code: String, completionHandler handler: @escaping LRAPIResponseHandler) {
        LoginRadiusREST.apiInstance.sendGET(url: "api/v2/access_token/wechat", queryParams: [
            "key": LoginRadiusSDK.sharedInstance.apiKey,
            "code": code
        ]) { (data, error) in
            handler(data, error)
        }
    }
    
    /**
     Native Apple Login
     
     @param code Apple Access Code
     @param socialAppName  should have unique social app name as a provider in case of multiple social apps for the same provider (eg. apple_<social app name> )
     @param handler code block executed after completion
     */
    public func convertAppleCodeToLRToken( code: String, withSocialAppName socialAppName: String?, completionHandler handler: @escaping LRAPIResponseHandler) {
        var dictParam: [String: Any] = [
            "key": LoginRadiusSDK.sharedInstance.apiKey,
            "code": code
        ]
        
        if let socialAppName = socialAppName, !socialAppName.isEmpty {
            dictParam["socialappname"] = socialAppName
        }
        
        LoginRadiusREST.apiInstance.sendGET(url: "api/v2/access_token/apple", queryParams: dictParam) { (data, error) in
            handler(data, error)
        }
    }
    
    
    /**
     *  Log out the user
     */
    public func logout() {
        // Only facebook native login stores sessions that we have to clear.
        self.facebookLogin.logout()
    }
    
    
    // MARK: - AppDelegate methods
    
    public func applicationLaunched(withOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
        self.facebookLogin.applicationLaunched(withOptions: launchOptions)
    }
    
    /**
     *  Call this for native social login to work properly
     */
    public func application(_ application: UIApplication, open url: URL, sourceApplication: String?, annotation: Any) -> Bool {
        var canOpen = false
        
        if self.isFacebookNativeLogin {
            self.isFacebookNativeLogin = false
            canOpen = self.facebookLogin.application(application, open: url, sourceApplication: sourceApplication, annotation: annotation)
        } else if self.isSafariLogin {
            self.isSafariLogin = false
            canOpen = self.safariLogin.application(application, open: url, sourceApplication: sourceApplication, annotation: annotation)
        }
        
        return canOpen
    }
}

